File: typescript/mirascope/src/llm/calls/async-call.ts
/**
 * @fileoverview The AsyncCall module for generating responses asynchronously using LLMs.
 */

import type { AsyncPrompt } from "../prompts";
import type { Response } from "../responses";
import { BaseCall } from "./base-call";

/**
 * A class for generating responses using LLMs asynchronously.
 */
export class AsyncCall<P extends any[] = any[]> extends BaseCall<P, AsyncPrompt<P>> {
  // Overloads for spread parameters
  call(...args: P): Promise<Response>;
  
  // Overload for single params object
  call(params: P extends [infer T] ? T : never): Promise<Response>;
  
  /**
   * Generates a response using the LLM asynchronously.
   * Accepts either spread parameters or a single params object.
   */
  async call(...args: any[]): Promise<Response> {
    throw new Error("Not implemented");
  }

  /**
   * Generates an asynchronous response using the LLM.
   * Alias for the call method.
   */
  async callAsync(...args: P): Promise<Response> {
    return await this.call(...args);
  }
}

File: python/mirascope/llm/calls/async_call.py
"""The AsyncCall module for generating responses asynchronously using LLMs."""

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Generic

from ..content import UserContent
from ..prompts import AsyncPrompt
from ..response_formatting import FormatT
from ..responses import Response
from ..streams import AsyncStream, BaseStream
from ..tools import Toolkit, ToolT
from ..types import P
from .base_call import BaseCall


@dataclass
class AsyncCall(BaseCall[P, AsyncPrompt, FormatT], Generic[P, ToolT, FormatT]):
    """A class for generating responses using LLMs asynchronously."""

    toolkit: Toolkit[ToolT]
    """The toolkit of tools associated with this call."""

    async def __call__(
        self, *args: P.args, **kwargs: P.kwargs
    ) -> Response[None, FormatT]:
        """Generates a response using the LLM asynchronously."""
        raise NotImplementedError()

    async def call(self, *args: P.args, **kwargs: P.kwargs) -> Response[None, FormatT]:
        """Generates a response using the LLM asynchronously."""
        raise NotImplementedError()

    async def call_async(
        self, *args: P.args, **kwargs: P.kwargs
    ) -> Response[None, FormatT]:
        """Generates an asynchronous response using the LLM."""
        raise NotImplementedError()

    async def stream(
        self, *args: P.args, **kwargs: P.kwargs
    ) -> AsyncStream[None, FormatT]:
        """Generates a streaming response using the LLM asynchronously."""
        raise NotImplementedError()

    async def stream_async(
        self, *args: P.args, **kwargs: P.kwargs
    ) -> AsyncStream[None, FormatT]:
        """Generates an asynchronous streaming response using the LLM."""
        raise NotImplementedError()

    async def resume(
        self,
        output: Response[None, FormatT] | BaseStream[None, FormatT],
        content: UserContent | Sequence[UserContent],
    ) -> Response[None, FormatT]:
        """Generate a new response by continuing from a previous output, plus new user content."""
        raise NotImplementedError()

    async def resume_async(
        self,
        output: Response[None, FormatT] | BaseStream[None, FormatT],
        content: UserContent | Sequence[UserContent],
    ) -> Response[None, FormatT]:
        """Generate a new response asynchronously by continuing from a previous output, plus new user content."""
        raise NotImplementedError()

    async def resume_stream(
        self,
        output: Response[None, FormatT] | BaseStream[None, FormatT],
        content: UserContent | Sequence[UserContent],
    ) -> AsyncStream[None, FormatT]:
        """Generate a new stream by continuing from a previous output, plus new user content."""
        raise NotImplementedError()

    async def resume_stream_async(
        self,
        output: Response[None, FormatT] | BaseStream[None, FormatT],
        content: UserContent | Sequence[UserContent],
    ) -> AsyncStream[None, FormatT]:
        """Generate a new async stream by continuing from a previous output, plus new user content."""
        raise NotImplementedError()
